// @dracul/settings-backend/lib/models/SettingsModel.js
"use strict";

var _commonBackend = require("@dracul/common-backend");

var mongoosePaginate = require('mongoose-paginate-v2');

var uniqueValidator = require('mongoose-unique-validator');

var Schema = _commonBackend.mongoose.Schema;
var SettingsSchema = new Schema({
  key: {
    type: String,
    required: true,
    unique: true
  },
  value: {
    type: String,
    required: false,
    unique: false
  },
  label: {
    en: {
      type: String,
      required: false
    },
    es: {
      type: String,
      required: false
    },
    pt: {
      type: String,
      required: false
    }
  },
  type: {
    type: String,
    "default": "string",
    "enum": ['string', 'number', 'enum', 'boolean'],
    required: false,
    unique: false
  },
  options: [{
    type: String
  }]
});
SettingsSchema.plugin(mongoosePaginate);
SettingsSchema.plugin(uniqueValidator, {
  message: 'validation.unique'
});

var Settings = _commonBackend.mongoose.model('Settings', SettingsSchema);

module.exports = Settings;
