// @dracul/user-backend/lib/graphql/resolvers/AuthResolvers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _UserService = require("../../services/UserService");

var _AuthService = require("../../services/AuthService");

var _apolloServerExpress = require("apollo-server-express");

var _permissions = require("../../permissions");

var _default = {
  Query: {
    me: (_, {}, {
      user
    }) => {
      if (!user) throw new _apolloServerExpress.AuthenticationError("UNAUTHENTICATED");
      return (0, _UserService.findUser)(user.id);
    }
  },
  Mutation: {
    auth: (_, {
      username,
      password
    }, {
      req
    }) => {
      return new Promise((resolve, reject) => {
        (0, _AuthService.auth)({
          username,
          password
        }, req).then(r => resolve({
          token: r.token,
          refreshToken: r.refreshToken
        })).catch(err => {
          console.warn('Auth error: ', err.message);
          reject(new _apolloServerExpress.AuthenticationError("BadCredentials"));
        });
      });
    },
    apikey: (_, {
      userid
    }, {
      user,
      rbac,
      req
    }) => {
      if (!user) throw new _apolloServerExpress.AuthenticationError("UNAUTHENTICATED");
      if (!user || !rbac.isAllowed(user.id, _permissions.SECURITY_USER_EDIT)) throw new _apolloServerExpress.ForbiddenError("Not Authorized");
      return new Promise((resolve, reject) => {
        (0, _AuthService.apiKey)(userid, req).then(r => resolve({
          token: r.token
        })).catch(err => {
          console.warn('ApiKey error: ', err.message);
          reject(new _apolloServerExpress.AuthenticationError("BadCredentials"));
        });
      });
    },
    refreshToken: (_, {
      refreshTokenId
    }, {
      req
    }) => {
      return new Promise((resolve, reject) => {
        (0, _AuthService.refreshAuth)(refreshTokenId).then(r => resolve({
          token: r
        })).catch(err => {
          //console.warn('Auth error: ', err.message)
          reject(err);
        });
      });
    }
  }
};
exports.default = _default;
