// @dracul/user-backend/lib/graphql/resolvers/RoleResolvers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _RoleService = require("../../services/RoleService");

var _permissions = require("../../permissions");

var _apolloServerExpress = require("apollo-server-express");

var _UserService = require("../../services/UserService");

var _default = {
  Query: {
    roles: async (_, {
      id
    }, {
      user,
      rbac
    }) => {
      if (!user) throw new _apolloServerExpress.AuthenticationError("Unauthenticated");
      user = await (0, _UserService.findUser)(user.id);

      if (rbac.isAllowed(user.id, _permissions.SECURITY_ROLE_SHOW)) {
        return (0, _RoleService.findRoles)();
      } else if (rbac.isAllowed(user.id, _permissions.SECURITY_ROLE_SHOW_CHILD)) {
        return (0, _RoleService.findRoles)(user.role.childRoles);
      } else {
        throw new _apolloServerExpress.ForbiddenError("Not Authorized");
      }
    },
    role: (_, {
      id
    }, {
      user,
      rbac
    }) => {
      if (!user) throw new _apolloServerExpress.AuthenticationError("Unauthenticated");
      if (!user || !rbac.isAllowed(user.id, _permissions.SECURITY_ROLE_SHOW)) throw new _apolloServerExpress.ForbiddenError("Not Authorized");
      return (0, _RoleService.findRole)(id);
    }
  },
  Mutation: {
    roleCreate: (_, {
      input
    }, {
      user,
      rbac
    }) => {
      if (!user) throw new _apolloServerExpress.AuthenticationError("Unauthenticated");
      if (!user || !rbac.isAllowed(user.id, _permissions.SECURITY_ROLE_CREATE)) throw new _apolloServerExpress.ForbiddenError("Not Authorized");
      return (0, _RoleService.createRole)(input);
    },
    roleUpdate: (_, {
      id,
      input
    }, {
      user,
      rbac
    }) => {
      if (!user) throw new _apolloServerExpress.AuthenticationError("Unauthenticated");
      if (!rbac.isAllowed(user.id, _permissions.SECURITY_ROLE_EDIT)) throw new _apolloServerExpress.ForbiddenError("Not Authorized");
      return (0, _RoleService.updateRole)(id, input);
    },
    roleDelete: (_, {
      id
    }, {
      user,
      rbac
    }) => {
      if (!user) throw new _apolloServerExpress.AuthenticationError("Unauthenticated");
      if (!rbac.isAllowed(user.id, _permissions.SECURITY_ROLE_DELETE)) throw new _apolloServerExpress.ForbiddenError("Not Authorized");
      return (0, _RoleService.deleteRole)(id);
    }
  }
};
exports.default = _default;
