// @dracul/user-backend/lib/services/PermissionService.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updatePermission = exports.findPermissionByName = exports.findPermission = exports.fetchPermissionsInName = exports.fetchPermissions = exports.deletePermission = exports.createPermission = void 0;

var _PermissionModel = _interopRequireDefault(require("../models/PermissionModel"));

var _apolloServerExpress = require("apollo-server-express");

var _loggerBackend = require("@dracul/logger-backend");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const fetchPermissions = function () {
  return new Promise((resolve, reject) => {
    _PermissionModel.default.find({}).exec((err, res) => {
      if (err) {
        _loggerBackend.DefaultLogger.error("PermissionService.fetchPermissions ", err);

        reject(err);
      }

      resolve(res);
    });
  });
};

exports.fetchPermissions = fetchPermissions;

const fetchPermissionsInName = function (permissions) {
  return new Promise((resolve, reject) => {
    _PermissionModel.default.find({
      name: {
        $in: permissions
      }
    }).exec((err, res) => {
      if (err) {
        _loggerBackend.DefaultLogger.error("PermissionService.fetchPermissionsInName ", err);

        reject(err);
      }

      resolve(res);
    });
  });
};

exports.fetchPermissionsInName = fetchPermissionsInName;

const findPermission = function (id) {
  return new Promise((resolve, reject) => {
    _PermissionModel.default.findOne({
      _id: id
    }).exec((err, res) => {
      if (err) {
        _loggerBackend.DefaultLogger.error("PermissionService.findPermission ", err);

        reject(err);
      }

      resolve(res);
    });
  });
};

exports.findPermission = findPermission;

const findPermissionByName = function (name) {
  return new Promise((resolve, reject) => {
    _PermissionModel.default.findOne({
      name: name
    }).exec((err, res) => {
      if (err) {
        _loggerBackend.DefaultLogger.error("PermissionService.findPermission ", err);

        reject(err);
      }

      resolve(res);
    });
  });
};

exports.findPermissionByName = findPermissionByName;

const createPermission = function (name) {
  const newPermission = new _PermissionModel.default({
    name
  });
  newPermission.id = newPermission._id;
  return new Promise((resolve, reject) => {
    newPermission.save(err => {
      if (err) {
        _loggerBackend.DefaultLogger.error("PermissionService.createPermission ", err);

        reject(err);
      }

      resolve(newPermission);
    });
  });
};

exports.createPermission = createPermission;

const updatePermission = async function (id, name) {
  return new Promise((resolve, reject) => {
    _PermissionModel.default.findOneAndUpdate({
      _id: id
    }, {
      name,
      permissions
    }, {
      new: true,
      runValidators: true,
      context: 'query'
    }, (error, doc) => {
      if (error) {
        if (error.name == "ValidationError") {
          _loggerBackend.DefaultLogger.warn("PermissionService.updatePermission.ValidationError ", err);

          rejects(new _apolloServerExpress.UserInputError(error.message, {
            inputErrors: error.errors
          }));
        }

        _loggerBackend.DefaultLogger.error("PermissionService.updatePermission ", err);

        reject(error);
      }

      resolve(doc);
    });
  });
};

exports.updatePermission = updatePermission;

const deletePermission = function (id) {
  return new Promise((resolve, reject) => {
    findPermission(id).then(doc => {
      doc.softdelete(function (err) {
        if (err) {
          _loggerBackend.DefaultLogger.error("PermissionService.deletePermission ", err);

          reject(err);
        }

        resolve({
          id: id,
          success: true
        });
      });
    });
  });
};

exports.deletePermission = deletePermission;
