// @dracul/user-backend/lib/services/RegisterService.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerUser = exports.activationUser = void 0;

var _loggerBackend = require("@dracul/logger-backend");

var _RoleService = require("./RoleService");

var _UserModel = _interopRequireDefault(require("../models/UserModel"));

var _apolloServerExpress = require("apollo-server-express");

var _jsonwebtoken = _interopRequireDefault(require("jsonwebtoken"));

var _UserAuditService = require("./UserAuditService");

var _UserEmailManager = _interopRequireDefault(require("./UserEmailManager"));

var _UserService = require("./UserService");

var _AuthService = require("./AuthService");

var _SessionService = require("./SessionService");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const registerUser = function ({
  username,
  password,
  name,
  email,
  phone
}) {
  console.log(`DATA!: '${(username, password, name, email, phone)}'`);
  return new Promise(async (resolve, reject) => {
    const ROLE_NAME = process.env.REGISTER_ROLE ? process.env.REGISTER_ROLE : "operator";
    let roleObject = await (0, _RoleService.findRoleByName)(ROLE_NAME);
    let active = false;
    const newUser = new _UserModel.default({
      username,
      email,
      password: (0, _UserService.hashPassword)(password),
      name,
      phone,
      active,
      role: roleObject,
      createdAt: Date.now()
    });
    newUser.id = newUser._id;
    newUser.save(error => {
      if (error) {
        if (error.name == "ValidationError") {
          _loggerBackend.DefaultLogger.warn("RegisterService.registerUser.ValidationError ", error);

          reject(new _apolloServerExpress.UserInputError(error.message, {
            inputErrors: error.errors
          }));
        } else {
          _loggerBackend.DefaultLogger.error("RegisterService.registerUser ", error);
        }

        reject(error);
      } else {
        let token = _jsonwebtoken.default.sign({
          id: newUser.id,
          operation: 'register'
        }, process.env.JWT_SECRET, {
          expiresIn: process.env.JWT_REGISTER_EXPIRED_IN || '30d'
        });

        const url = `${process.env.APP_WEB_URL}/activation/${token}`;
        (0, _UserAuditService.createUserAudit)(newUser.id, newUser.id, 'userRegistered');

        _UserEmailManager.default.activation(newUser.email, url, newUser);

        _loggerBackend.DefaultLogger.info("RegisterService.registerUser successful for " + newUser.username);

        resolve({
          status: true,
          id: newUser.id,
          email: newUser.email
        });
      }
    });
  });
};

exports.registerUser = registerUser;

const activationUser = function (token, req) {
  return new Promise((resolve, reject) => {
    const decoded = _jsonwebtoken.default.verify(token, process.env.JWT_SECRET);

    if (!decoded) resolve({
      status: false,
      message: "common.operation.fail"
    });

    _UserModel.default.findOneAndUpdate({
      _id: decoded.id
    }, {
      active: true
    }).populate('role').populate('groups').exec((error, user) => {
      if (error) {
        _loggerBackend.DefaultLogger.error("RegisterService.activationUser.findOneAndUpdate ", error);

        resolve({
          status: false,
          message: "common.operation.fail"
        });
      }

      (0, _UserAuditService.createUserAudit)(user._id, user._id, 'userActivated');
      (0, _SessionService.createSession)(user, req).then(session => {
        const payload = (0, _AuthService.tokenSignPayload)(user, session.id);
        const options = {
          expiresIn: process.env.JWT_LOGIN_EXPIRED_IN || '1d',
          jwtid: user.id
        };

        let token = _jsonwebtoken.default.sign(payload, process.env.JWT_SECRET, options);

        _loggerBackend.DefaultLogger.info("RegisterService.activationUser successful for " + user.username);

        resolve({
          status: true,
          token: token,
          message: "common.operation.success"
        });
      }).catch(error => {
        _loggerBackend.DefaultLogger.error("RegisterService.activationUser ", error);

        reject(error);
      });
    });
  });
};

exports.activationUser = activationUser;
