// @dracul/user-backend/lib/services/RoleService.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateRole = exports.findRoles = exports.findRoleByName = exports.findRole = exports.fetchRolesInName = exports.deleteRole = exports.createRole = void 0;

var _loggerBackend = require("@dracul/logger-backend");

var _RoleModel = _interopRequireDefault(require("../models/RoleModel"));

var _apolloServerExpress = require("apollo-server-express");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const fetchRolesInName = function (roleNames) {
  return new Promise((resolve, reject) => {
    _RoleModel.default.find({
      name: {
        $in: roleNames
      }
    }).exec((err, res) => {
      if (err) {
        _loggerBackend.DefaultLogger.error("RoleService.fetchRolesInName ", err);

        reject(err);
      }

      _loggerBackend.DefaultLogger.debug('RoleService.fetchRolesInName successful');

      resolve(res);
    });
  });
};

exports.fetchRolesInName = fetchRolesInName;

const findRoles = function (roles = []) {
  return new Promise((resolve, reject) => {
    let qs = {};

    if (roles && roles.length) {
      qs._id = {
        $in: roles
      };
    }

    _RoleModel.default.find(qs).isDeleted(false).exec((err, res) => {
      if (err) {
        _loggerBackend.DefaultLogger.error("RoleService.findRoles ", err);

        reject(err);
      }

      _loggerBackend.DefaultLogger.debug('RoleService.findRoles successful');

      resolve(res);
    });
  });
};

exports.findRoles = findRoles;

const findRole = function (id) {
  return new Promise((resolve, reject) => {
    _RoleModel.default.findOne({
      _id: id
    }).exec((err, res) => {
      if (err) {
        _loggerBackend.DefaultLogger.error("RoleService.findRole ", err);

        reject(err);
      }

      _loggerBackend.DefaultLogger.debug('RoleService.findRole successful');

      resolve(res);
    });
  });
};

exports.findRole = findRole;

const findRoleByName = function (roleName) {
  return new Promise((resolve, reject) => {
    _RoleModel.default.findOne({
      name: roleName
    }).exec((err, res) => {
      if (err) {
        _loggerBackend.DefaultLogger.error("RoleService.findRoleByName ", err);

        reject(err);
      }

      _loggerBackend.DefaultLogger.debug('RoleService.findRoleByName successful');

      resolve(res);
    });
  });
};

exports.findRoleByName = findRoleByName;

const deleteRole = function (id) {
  return new Promise((resolve, rejects) => {
    findRole(id).then(doc => {
      doc.softdelete(function (err) {
        if (err) {
          _loggerBackend.DefaultLogger.error("RoleService.deleteRole ", err);

          reject(err);
        }

        _loggerBackend.DefaultLogger.info('RoleService.deleteRole successful');

        resolve({
          id: id,
          success: true
        });
      });
    });
  });
};

exports.deleteRole = deleteRole;

const createRole = function ({
  name,
  childRoles,
  permissions,
  readonly
}) {
  const newRole = new _RoleModel.default({
    name,
    childRoles,
    permissions,
    readonly
  });
  newRole.id = newRole._id;
  return new Promise((resolve, rejects) => {
    newRole.save(error => {
      if (error) {
        if (error.name == "ValidationError") {
          _loggerBackend.DefaultLogger.warn("RoleService.createRole.ValidationError ", error);

          rejects(new _apolloServerExpress.UserInputError(error.message, {
            inputErrors: error.errors
          }));
        } else {
          _loggerBackend.DefaultLogger.error("RoleService.createRole ", error);
        }

        rejects(error);
      } else {
        _loggerBackend.DefaultLogger.info('RoleService.createRole successful');

        resolve(newRole);
      }
    });
  });
};

exports.createRole = createRole;

const updateRole = async function (id, {
  name,
  childRoles,
  permissions = [],
  readonly
}) {
  return new Promise((resolve, rejects) => {
    _RoleModel.default.findOneAndUpdate({
      _id: id
    }, {
      name,
      childRoles,
      permissions,
      readonly
    }, {
      new: true,
      runValidators: true,
      context: 'query'
    }, (error, doc) => {
      if (error) {
        if (error.name == "ValidationError") {
          _loggerBackend.DefaultLogger.warn("RoleService.updateRole.ValidationError ", error);

          rejects(new _apolloServerExpress.UserInputError(error.message, {
            inputErrors: error.errors
          }));
        } else {
          _loggerBackend.DefaultLogger.error("RoleService.updateRole ", error);
        }

        rejects(error);
      }

      _loggerBackend.DefaultLogger.info('RoleService.updateRole successful');

      resolve(doc);
    });
  });
};

exports.updateRole = updateRole;
