// @dracul/customize-backend/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createCustomization", {
  enumerable: true,
  get: function () {
    return _CustomizationService.createCustomization;
  }
});
Object.defineProperty(exports, "findCustomization", {
  enumerable: true,
  get: function () {
    return _CustomizationService.findCustomization;
  }
});
Object.defineProperty(exports, "initCustomization", {
  enumerable: true,
  get: function () {
    return _InitCustomizationService.initCustomization;
  }
});
Object.defineProperty(exports, "initPermissionsCustomization", {
  enumerable: true,
  get: function () {
    return _InitCustomizationService.initPermissionsCustomization;
  }
});
exports.permissions = void 0;
Object.defineProperty(exports, "resolvers", {
  enumerable: true,
  get: function () {
    return _graphql.resolvers;
  }
});
Object.defineProperty(exports, "types", {
  enumerable: true,
  get: function () {
    return _graphql.types;
  }
});
Object.defineProperty(exports, "updateColors", {
  enumerable: true,
  get: function () {
    return _CustomizationService.updateColors;
  }
});
Object.defineProperty(exports, "updateCustomization", {
  enumerable: true,
  get: function () {
    return _CustomizationService.updateCustomization;
  }
});
Object.defineProperty(exports, "updateLang", {
  enumerable: true,
  get: function () {
    return _CustomizationService.updateLang;
  }
});
Object.defineProperty(exports, "updateLogo", {
  enumerable: true,
  get: function () {
    return _CustomizationService.updateLogo;
  }
});
Object.defineProperty(exports, "uploadLogo", {
  enumerable: true,
  get: function () {
    return _CustomizationService.uploadLogo;
  }
});

var _CustomizationService = require("./services/CustomizationService");

var permissions = _interopRequireWildcard(require("./permissions"));

exports.permissions = permissions;

var _graphql = require("./graphql");

var _InitCustomizationService = require("./services/InitCustomizationService");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
