// @dracul/settings-backend/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "SettingCache", {
  enumerable: true,
  get: function get() {
    return _SettingCache["default"];
  }
});
Object.defineProperty(exports, "SettingsModel", {
  enumerable: true,
  get: function get() {
    return _SettingsModel["default"];
  }
});
Object.defineProperty(exports, "createSettings", {
  enumerable: true,
  get: function get() {
    return _SettingsService.createSettings;
  }
});
Object.defineProperty(exports, "deleteSettings", {
  enumerable: true,
  get: function get() {
    return _SettingsService.deleteSettings;
  }
});
Object.defineProperty(exports, "fetchSettings", {
  enumerable: true,
  get: function get() {
    return _SettingsService.fetchSettings;
  }
});
Object.defineProperty(exports, "findSettings", {
  enumerable: true,
  get: function get() {
    return _SettingsService.findSettings;
  }
});
Object.defineProperty(exports, "findSettingsByKey", {
  enumerable: true,
  get: function get() {
    return _SettingsService.findSettingsByKey;
  }
});
Object.defineProperty(exports, "findSettingsByKeys", {
  enumerable: true,
  get: function get() {
    return _SettingsService.findSettingsByKeys;
  }
});
Object.defineProperty(exports, "getSettingsValueByKey", {
  enumerable: true,
  get: function get() {
    return _SettingsService.getSettingsValueByKey;
  }
});
Object.defineProperty(exports, "initializeSetting", {
  enumerable: true,
  get: function get() {
    return _SettingsService.initializeSetting;
  }
});
Object.defineProperty(exports, "initializeSettings", {
  enumerable: true,
  get: function get() {
    return _SettingsService.initializeSettings;
  }
});
Object.defineProperty(exports, "paginateSettings", {
  enumerable: true,
  get: function get() {
    return _SettingsService.paginateSettings;
  }
});
exports.permissions = void 0;
Object.defineProperty(exports, "resolvers", {
  enumerable: true,
  get: function get() {
    return _graphql.resolvers;
  }
});
Object.defineProperty(exports, "types", {
  enumerable: true,
  get: function get() {
    return _graphql.types;
  }
});
Object.defineProperty(exports, "updateSettings", {
  enumerable: true,
  get: function get() {
    return _SettingsService.updateSettings;
  }
});
Object.defineProperty(exports, "updateSettingsByKey", {
  enumerable: true,
  get: function get() {
    return _SettingsService.updateSettingsByKey;
  }
});

var _graphql = require("./graphql");

var _SettingsModel = _interopRequireDefault(require("./models/SettingsModel"));

var permissions = _interopRequireWildcard(require("./permissions/Settings"));

exports.permissions = permissions;

var _SettingsService = require("./services/SettingsService");

var _SettingCache = _interopRequireDefault(require("./cache/SettingCache"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
