// @dracul/user-backend/lib/models/GroupModel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.GroupSchema = exports.GroupModel = void 0;

var _commonBackend = require("@dracul/common-backend");

var _mongooseSoftdelete = _interopRequireDefault(require("mongoose-softdelete"));

var _mongoosePaginateV = _interopRequireDefault(require("mongoose-paginate-v2"));

var _mongooseUniqueValidator = _interopRequireDefault(require("mongoose-unique-validator"));

var _GroupService = require("../services/GroupService");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const Schema = _commonBackend.mongoose.Schema;
const GroupSchema = new Schema({
  name: {
    type: String,
    required: true,
    unique: false,
    index: false
  },
  color: {
    type: String,
    required: false,
    validate: {
      validator: function (value) {
        let r = /^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/;
        return r.test(value);
      },
      message: "validation.invalidHexColor"
    }
  },
  users: [{
    type: _commonBackend.mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: false
  }]
});
exports.GroupSchema = GroupSchema;
GroupSchema.plugin(_mongooseUniqueValidator.default, {
  message: 'validation.unique'
});
GroupSchema.plugin(_mongooseSoftdelete.default);
GroupSchema.plugin(_mongoosePaginateV.default);
GroupSchema.set('toJSON', {
  transform: (doc, result) => {
    return { ...result,
      id: result._id
    };
  }
});

const GroupModel = _commonBackend.mongoose.model('Group', GroupSchema);

exports.GroupModel = GroupModel;
var _default = GroupModel; //module.exports = Group;

exports.default = _default;
