// @dracul/user-backend/lib/models/RoleModel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.RoleSchema = exports.RoleModel = void 0;

var _commonBackend = require("@dracul/common-backend");

var _mongooseSoftdelete = _interopRequireDefault(require("mongoose-softdelete"));

var _mongoosePaginateV = _interopRequireDefault(require("mongoose-paginate-v2"));

var _mongooseUniqueValidator = _interopRequireDefault(require("mongoose-unique-validator"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const Schema = _commonBackend.mongoose.Schema;
const RoleSchema = new Schema({
  name: {
    type: String,
    unique: true,
    required: true,
    index: true
  },
  permissions: [{
    type: String,
    required: true
  }],
  childRoles: [{
    type: _commonBackend.mongoose.Schema.Types.ObjectId,
    ref: 'Role',
    required: false
  }],
  readonly: {
    type: Boolean,
    required: false,
    default: false
  }
});
exports.RoleSchema = RoleSchema;
RoleSchema.plugin(_mongooseUniqueValidator.default, {
  message: 'validation.unique'
});
RoleSchema.plugin(_mongooseSoftdelete.default);
RoleSchema.plugin(_mongoosePaginateV.default);
RoleSchema.set('toJSON', {
  getters: true
});

const RoleModel = _commonBackend.mongoose.model('Role', RoleSchema);

exports.RoleModel = RoleModel;
var _default = RoleModel; //module.exports = RoleModel

exports.default = _default;
