// @dracul/user-backend/lib/models/UserModel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.UserSchema = exports.UserModel = void 0;

var _commonBackend = require("@dracul/common-backend");

var _mongooseSoftdelete = _interopRequireDefault(require("mongoose-softdelete"));

var _mongoosePaginateV = _interopRequireDefault(require("mongoose-paginate-v2"));

var _mongooseUniqueValidator = _interopRequireDefault(require("mongoose-unique-validator"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Defining user Mongoose Schema
const UserSchema = new _commonBackend.mongoose.Schema({
  username: {
    type: String,
    unique: true,
    required: true,
    index: true
  },
  email: {
    type: String,
    unique: true,
    required: true,
    index: true,
    validate: {
      validator: function (value) {
        let r = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
        return r.test(value);
      },
      message: "validation.emailFormat"
    }
  },
  password: {
    type: String,
    required: true
  },
  code: {
    type: String,
    required: false
  },
  name: {
    type: String,
    required: true
  },
  active: {
    type: Boolean,
    required: true,
    default: false
  },
  phone: {
    type: String,
    required: false,
    validate: {
      validator: function (value) {
        let r = /[0-9]+/;
        return value ? r.test(value) : true;
      },
      message: "Telefono no tiene un formato valido"
    }
  },
  avatar: String,
  avatarurl: String,
  role: {
    type: _commonBackend.mongoose.Schema.Types.ObjectId,
    ref: 'Role',
    required: true
  },
  groups: [{
    type: _commonBackend.mongoose.Schema.Types.ObjectId,
    ref: 'Group',
    required: false
  }],
  refreshToken: {
    type: [{
      id: {
        type: String
      },
      expiryDate: {
        type: Date
      },
      sessionId: {
        type: String
      }
    }],
    default: [],
    _id: false
  },
  fromLDAP: {
    type: Boolean,
    required: true,
    default: false
  }
}, {
  timestamps: true
});
exports.UserSchema = UserSchema;
UserSchema.set('toJSON', {
  getters: true
});
UserSchema.plugin(_mongooseUniqueValidator.default, {
  message: 'validation.unique'
});
UserSchema.plugin(_mongooseSoftdelete.default);
UserSchema.plugin(_mongoosePaginateV.default);

const UserModel = _commonBackend.mongoose.model('User', UserSchema);

exports.UserModel = UserModel;
var _default = UserModel;
exports.default = _default;
