// @dracul/customize-backend/lib/services/InitCustomizationService.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.initPermissionsCustomization = exports.initCustomization = void 0;

var _loggerBackend = require("@dracul/logger-backend");

var _permissions = require("../permissions");

var _userBackend = require("@dracul/user-backend");

var _CustomizationService = require("./CustomizationService");

const initCustomization = async function () {
  let customDoc = await (0, _CustomizationService.findCustomization)();

  if (!customDoc) {
    let customDoc = await (0, _CustomizationService.createCustomization)({
      colors: {
        primary: '#3F51B5',
        onPrimary: '#FFFFFF',
        secondary: '#1565C0',
        onSecondary: '#FFFFFF'
      },
      logo: {
        mode: 'OnlyTitle',
        title: 'APP'
      },
      language: 'en'
    });

    _loggerBackend.DefaultLogger.info("customization created: ", customDoc.id);
  } else {
    _loggerBackend.DefaultLogger.debug("customization found: ", customDoc.id);
  }
};

exports.initCustomization = initCustomization;

const initPermissionsCustomization = async function () {
  let permissions = [_permissions.CUSTOMIZATION_SHOW, _permissions.CUSTOMIZATION_CREATE, _permissions.CUSTOMIZATION_UPDATE, _permissions.CUSTOMIZATION_COLORS_UPDATE, _permissions.CUSTOMIZATION_LANG_UPDATE, _permissions.CUSTOMIZATION_LOGO_UPDATE];
  await _userBackend.InitService.initPermissions(permissions);
};

exports.initPermissionsCustomization = initPermissionsCustomization;
